### DISCUSSION TITLE
q-form validation not working as expected

### DISCUSSION THREAD

**Discussion by jnui** [quasarframework/quasar#9778] [General - Components / Directives / etc] [answered]
I'm on vue 2, and trying to validate form fields.
The rules work on the input fields e.g. if I try to tab away from a required field I get the validation failed message.
I want to be able to click submit, and all fields to be validated pass/fail and the form to not be submitted.
2 problems.
1. if I use q-form and q-btn type=submit then the @submit function is not fired.
2. if I use q-btn @click , then the function is fired, but it does not validate correctly (seems to always pass) but does not highlight field

how is this supposed to work?

```
<q-form
      @submit="submitUpdateJob"
      ref="jobForm"
      />
<q-input
               outlined
               v-model=job.zip
               label="ZIP"
               lazy-rules
               class="col"
               :rules="[val => !!val || 'Field is required']"
            />
         </div>
         <div align="right">
               <q-btn flat label="SUBMIT" type="submit"  color="primary"  />
               <!-- <q-btn flat label="CANCEL" color="red" @click="editMode = false" /> -->
         </div>
</q-form>
or...
<q-form
      ref="jobForm"
      />
 <q-input
               outlined
               v-model=job.zip
               label="ZIP"
               lazy-rules
               class="col"
               :rules="[val => !!val || 'Field is required']"
            />
         </div>
<div align="right">
     <q-btn flat label="SUBMIT" @click="submitUpdateJob"  color="primary"  />
</div>
</q-form>
...
methods: {
      submitUpdateJob: function () {
         this.$refs.jobForm.validate().then(success => {
            if(success){
              console.log('supposedly validated');
            } else {
               console.log('no success');
            }
         })
      },
```

**Comment by rstoenescu** [comment]
You don't need to trigger the validation by yourself on the `@submit` handler.
The `@submit` handler is only called if the validation succeeded.

**Reply by rstoenescu** [reply]
You'll have a better chance for getting help with a codepen/jsfiddle.
But this example in docs should explain it all: https://quasar.dev/vue-components/form

**Reply by jnui** [reply]
I forgot to mention that this form in a modal/dialog which seems to be causing the problem. (as mentioned on this bug https://github.com/quasarframework/quasar/issues/9450 which was supposed to be fixed on ver 2)
I got it to work by using your fix on that bug and adding a ref to each input, then running validate on each input by using the ref.
pseudo code...
```
<q-input
    outlined
    v-model=job.name
    label="Job Name"
    lazy-rules
    ref="jobName"
    :rules="[val => !!val || 'Field is required']"
 />
<q-btn @click="submitFormFunc" />
...
methods: {
   submitFormFunc: function(){
      if(this.$refs.jobName.validate()){
         //if pass validation for this field do something
      }
   }
}
```

**Comment by jnui** [comment, marked as answer]
answered

**Comment by VitalickS** [comment]
the answer is incorrect. The problem still exists
